example: variable inlining

obmarg opened the pull request:
TODO: 

- [ ] Commit message
- [ ] Document in the book
- [ ] Possibly build an actual shopify example
- [ ] Use `to_input_literal` to make it a bit smarter
- [ ] Possibly add functionality to the `QueryVariables` derive that makes `to_input_literal` easier to use?  Not sure.

Bobrokus commented:
Dang. You've put a lot of effort into this. Man, you don't even know how thankful I am. You even linked to the actual API reference.

Bobrokus commented:
I came up with this:

```rust
fn serialize_with_inline_arguments<Fragment, Vars>(
    operation: Operation<Fragment, Vars>,
) -> anyhow::Result<String>
where
    Fragment: cynic::QueryFragment,
    Vars: cynic::QueryVariables + serde::Serialize,
{
    // Serialize the operation into a Json Object
    let operation_value = serde_json::to_value(operation)?;
    let operation_map = operation_value
        .as_object()
        .context("Value is not an Object")?;

    // Get the query field from the operation Object
    let query = operation_map
        .get("variables")
        .context("Operation variables missing")?
        .as_object()
        .context("Value is not an Object")?
        .into_iter()
        .map(|(name, value)| {
            Ok((name.to_owned(), serde_json::to_string(value)?))
        })
        .collect::<anyhow::Result<Vec<(String, String)>>>()?
        .into_iter()
        .fold(
            operation_map
                .get("query")
                .context("Operation query string missing")?
                .as_str()
                .context("Value is not a String")?
                .to_string(),
            |query, (name, value)| query.replace(&format!("${name}"), &value),
        );

    Ok(query)
}
```
```rust
fn main() {
    let query = serialize_with_inline_arguments(StageUploads::build(
        StageUploadsVariables {
            input: vec![StagedUploadInput {
                resource: StagedUploadTargetGenerateUploadResource::File,
                filename: "whatever",
                mime_type: "revetahw",
                http_method: Some(StagedUploadHttpMethodType::Post),
                file_size: None,
            }],
        },
    ))
    .unwrap();

    println!("{query}");
}
```

Output:

```graphql
mutation StageUploads([{"fileSize":null,"filename":"whatever","httpMethod":"POST","mimeType":"revetahw","resource":"FILE"}]: [StagedUploadInput!]!) {
  stagedUploadsCreate(input: [{"fileSize":null,"filename":"whatever","httpMethod":"POST","mimeType":"revetahw","resource":"FILE"}]) {
    stagedTargets {
      resourceUrl
      url
      parameters {
        name
        value
      }
    }
    userErrors {
      field
      message
    }
  }
}
```

Of course, the variables in the declaration got replaced too but I'll be fixing that tomorrow. Important is that it replaced the arguments inside the query.

Bobrokus commented:
In [this article](https://learning.atheros.ai/blog/graphql-quick-tip-how-to-pass-variables-in-graphiql) they call it "Inline Fragment". I guess it doesn't really matter.

obmarg commented:
Yeah, I would say what they're doing in that article is putting arguments inline.  But in cynic terms I think it's fair to call it variable inlining - since we're taking variables and inlining them into arguments.

obmarg commented:
Writing out this example made me realise there were a few shortcomings with this approach (For more detail see the message in #1009 for more details).  I've now fixed those shortcomings and added this as a built in feature in #1012.

I want to keep this PR around as a record of this idea, so closing this in favour of #1013 which adds an example of the new approach.
